### ISSUE TITLE
Provide a Python API

### ISSUE THREAD

**Issue by djvulee** [BurntSushi/ripgrep#165]
The ripgrep is really awesome, If we can offer a Python API, maybe we can use ripgrep in a Wider area.

**Comment by BurntSushi** [comment]
While I agree this might be cool. Not even that, it should be eminently possible as well. [Someone is already working on a Python API for Rust's regex library.](https://github.com/davidblewett/regex/tree/rure-python/regex-capi/src/python)

With that said, the amount of work required to do this is quite a lot. It requires splitting everything out into a library, writing C interfaces and then writing idiomatic Python bindings. It's unlikely something I'll ever do personally, although I'd be happy to support such an effort.

I'd like to close this because this comes too far down on the side of wishlist as opposed to something that's likely to get worked on within the next *year*.

**Comment by h5rdly** [comment]
Hi,

I have no understanding of Rust, but does this bindings library perhaps allow for the task you described to be easier? It would be very useful to have a Python api to use rg

https://github.com/dgrunwald/rust-cpython

**Comment by BurntSushi** [comment]
@h5rdly I'm aware of that work and it doesn't necessarily make it easier. It depends on how you want to build the bindings. For example, if you use `cffi`, then I don't think there's any place for using something like `rust-cpython`. You can see how the regex bindings work here: https://github.com/davidblewett/rure-python/tree/master/rure

The hard work here is in crafting the C API, writing tests for it and then building a Pythonic interface to them.

**Comment by h5rdly** [comment]
That seems like a pain. I think bindings would allow a much wider use of rg, especially because it's cross platform already, but I can hardly press the matter :)

**Comment by BurntSushi** [comment]
Moreover, libripgrep isn't finished yet, so creating bindings to it today would be fruitless. 

Once libripgrep is done, this is a task i would be willing to mentor.

**Comment by BurntSushi** [comment]
In the mean time, consider a lower tech solution: running the rg executable and parsing the results. It wouldn't be too hard to write a library for it!

**Comment by h5rdly** [comment]
I figured the same, didn't want to say anything until I actually start :)

This is for a file manager I'm writing to teach myself python, and thought it'd be nice to have a powerful cross platform search tool as a backend. I'm just figuring out the basics, so this is a little down the road. 

I did put some thought into how I want to approach it, which is why I'm trying to convince @gpshead, the maintainer of subprocess32, to backport run(), which seems like a much more convenient function to work with than Popen() -
https://docs.python.org/3/library/subprocess.html#subprocess.run

so that such code is Python 2/3 compatible. Would be awesome if you could chime in.
https://github.com/google/python-subprocess32/issues/17

**Comment by honglei** [comment]
Adding a current data point in favor of reconsidering a Python API.

  I’m using Python to process large AIS text files and need to pre-filter an 11GB tab-separated file by MMSI before decoding matched AIS messages. For this workload, Python’s line-by-line `re` loop is much slower, while shelling out to `rg` works but is less ergonomic for library code.
  A small benchmark on the same 11GB file and regex gave consistent match counts:

  - `rg` CLI: ~2.7s warm
  - [ripgrep-rs](https://github.com/phil65/python-ripgrep): ~3.1s
  - [ripgrep-python](https://github.com/indent-com/python-ripgrep): ~3.1-3.4s
  - Python `re`: ~24-27s

  Given that PyO3/maturin and third-party bindings now exist, would it be worth reconsidering an official or semi-official Python binding, even with a small supported surface such as file search and matched line/count results?
